File: server/mcts.py
"""
Monte Carlo Tree Search implementation for Kalah
Optimized for GPU batch evaluation
"""

import numpy as np
import numpy.typing as npt
import math
from typing import Dict, Optional, Tuple, List
from collections import defaultdict
import threading
from kalah_game import KalahGame


class MCTSNode:
    """Node in the MCTS tree"""

    def __init__(self, prior: float = 0.0) -> None:
        """
        Initializes a new instance of the class.

        Args:
            prior (float, optional): The prior probability or value associated with this node. Defaults to 0.0.

        Attributes:
            visit_count (int): The number of times this node has been visited.
            value_sum (float): The cumulative value from all visits to this node.
            prior (float): The prior probability or value for this node.
            children (dict): A dictionary mapping actions to child nodes.
            virtual_loss (int): The virtual loss used for parallelization in MCTS.
        """
        self.visit_count = 0
        self.value_sum = 0.0
        self.prior = prior
        self.current_player = None
        self.children = {}
        self.virtual_loss = 0

    def value(self) -> float:
        """
        Calculates and returns the average value of the node.

        Returns:
            float: The average value (value_sum divided by visit_count) if visit_count is greater than 0,
                   otherwise 0.0.
        """
        if self.visit_count == 0:
            return 0.0
        return self.value_sum / self.visit_count

    def ucb_score(self, parent_visits: int, c_puct: float) -> float:
        """
        Calculates the Upper Confidence Bound (UCB) score for a node in Monte Carlo Tree Search (MCTS).

        The UCB score balances exploitation (the node's current value estimate) and
        exploration (the potential for discovering better outcomes). Unvisited nodes
        are prioritized by returning a high exploration value.

        Args:
            parent_visits (int): The total number of visits to the parent node.
            c_puct (float): The exploration constant that controls the balance between exploration and exploitation.

        Returns:
            float: The computed UCB score for the node.
        """
        if self.visit_count == 0:
            # Avoid division by zero, prioritize unvisited nodes
            return c_puct * self.prior * math.sqrt(parent_visits)

        exploitation = self.value()
        exploration = (
            c_puct * self.prior * math.sqrt(parent_visits) / (1 + self.visit_count)
        )

        return exploitation + exploration

    def is_expanded(self) -> bool:
        """
        Checks if the current node has been expanded.

        Returns:
            bool: True if the node has one or more child nodes, False otherwise.
        """
        return len(self.children) > 0


class MCTS:
    """
    Monte Carlo Tree Search for Kalah
    Optimized for GPU batch evaluation
    """

    def __init__(self, config, network, batch_size: int = 32):
        self.config = config
        self.network = network
        self.nodes = {}
        self.lock = threading.Lock()
        self.batch_size = batch_size

    def search(
        self, game: KalahGame, root_state: Optional[str] = None
    ) -> npt.NDArray[np.float64]:
        """
        Run MCTS simulations with batched neural network evaluation
        Args:
            game: KalahGame instance
            root_state: Optional root state key for tree reuse
        Returns:
            Action probabilities based on visit counts
        """
        if root_state is None:
            root_state = self._state_key(game)

        # Add Dirichlet noise to root node for exploration
        if root_state not in self.nodes:
            self._expand_node_single(game, root_state)

        root_node = self.nodes[root_state]
        root_node.current_player = game.current_player
        valid_moves = game.get_valid_moves()

        # Add Dirichlet noise to root prior
        noise = np.random.dirichlet(
            [self.config.mcts.dirichlet_alpha] * len(valid_moves)
        )
        for action in range(len(valid_moves)):
            if valid_moves[action] and action in root_node.children:
                child = root_node.children[action]
                child.prior = (
                    1 - self.config.mcts.dirichlet_epsilon
                ) * child.prior + self.config.mcts.dirichlet_epsilon * noise[action]

        # Collect leaf nodes in batches
        num_simulations = self.config.mcts.num_simulations
        simulation_count = 0

        while simulation_count < num_simulations:
            # Determine batch size for this iteration
            current_batch_size = min(
                self.batch_size, num_simulations - simulation_count
            )

            # Collect paths to leaf nodes
            leaf_infos = []
            for _ in range(current_batch_size):
                game_copy = game.clone()
                path, leaf_game, leaf_state_key = self._simulate_to_leaf(
                    game_copy, root_state
                )
                leaf_infos.append((path, leaf_game, leaf_state_key))

            # Batch evaluate all leaf nodes
            self._batch_evaluate_and_backup(leaf_infos)

            simulation_count += current_batch_size

        # Extract visit counts
        visits = np.zeros(6)
        for action, child in root_node.children.items():
            visits[action] = child.visit_count

        return visits

    def _simulate_to_leaf(
        self, game: KalahGame, root_state: str
    ) -> Tuple[List, KalahGame, str]:
        """
        Select down to a leaf node without evaluation
        Returns: (path, game_state, state_key)
        """
        path = []
        current_state = root_state

        # Selection phase - traverse tree until leaf
        while True:
            if current_state not in self.nodes:
                # Found unexpanded node
                return path, game, current_state

            node = self.nodes[current_state]
            node.current_player = game.current_player

            # Check if game is over
            if game.game_over:
                return path, game, current_state

            valid_moves = game.get_valid_moves()
            if not node.is_expanded():
                # Node exists but not expanded
                return path, game, current_state

            # Check if any valid moves exist
            if not np.any(valid_moves):
                return path, game, current_state

            action = self._select_action(node, valid_moves)
            if action is None:
                return path, game, current_state

            with self.lock:
                node.children[action].virtual_loss += 1

            path.append((current_state, action, game.current_player))

            # Make move and check for extra turn
            extra_turn = game.make_move(action)
            current_state = self._state_key(game)

            # Handle extra turns by continuing with same player context
            if extra_turn and not game.game_over:
                continue

            # Check game over after move
            if game.game_over:
                return path, game, current_state

    def _batch_evaluate_and_backup(
        self, leaf_infos: List[Tuple[List, KalahGame, str]]
    ) -> None:
        """
        Evaluate multiple leaf nodes in a single batch and backup values
        """
        # Separate terminal and non-terminal nodes
        non_terminal_infos = []
        terminal_values = []

        for i, (path, game, state_key) in enumerate(leaf_infos):
            if game.game_over:
                # Terminal node - use actual game outcome
                value = game.get_reward(game.current_player)
                terminal_values.append((i, value))
            else:
                non_terminal_infos.append((i, path, game, state_key))

        # Batch evaluate non-terminal nodes
        if non_terminal_infos:
            # Collect states for batch evaluation
            states = []
            for _, _, game, _ in non_terminal_infos:
                states.append(game.get_canonical_state())

            # Batch neural network evaluation
            states_array = np.array(states)
            policies, values = self.network.predict_batch(states_array)

            # Expand nodes with predicted policies
            for idx, (original_idx, path, game, state_key) in enumerate(
                non_terminal_infos
            ):
                if state_key not in self.nodes:
                    self._expand_node_with_policy(game, state_key, policies[idx])

                # Backup the value
                self._backup(path, values[idx])

        # Backup terminal values
        for original_idx, value in terminal_values:
            path = leaf_infos[original_idx][0]
            self._backup(path, value)

    def _expand_node_single(self, game: KalahGame, state_key: str) -> None:
        """
        Expand a single node (used for root node initialization)
        """
        if state_key in self.nodes:
            return

        # Get neural network predictions
        state = game.get_canonical_state()
        policy, value = self.network.predict(state)

        self._expand_node_with_policy(game, state_key, policy)

    def _expand_node_with_policy(
        self, game: KalahGame, state_key: str, policy: np.ndarray
    ) -> None:
        """
        Expand a node with a given policy
        """
        if state_key in self.nodes:
            return

        # Mask invalid actions and renormalize
        valid_moves = game.get_valid_moves()
        policy = policy * valid_moves
        policy_sum = np.sum(policy)

        if policy_sum > 0:
            policy = policy / policy_sum
        else:
            # All valid moves equally probable
            policy = valid_moves / np.sum(valid_moves)

        # Create node and children
        node = MCTSNode()
        for action in range(6):
            if valid_moves[action]:
                node.children[action] = MCTSNode(prior=policy[action])

        with self.lock:
            if state_key not in self.nodes:
                self.nodes[state_key] = node

    def _select_action(
        self, node: MCTSNode, valid_moves: npt.NDArray[np.bool_]
    ) -> Optional[int]:
        """
        Selects the best action from the given node based on the UCB (Upper Confidence Bound)
        score, considering only valid moves and accounting for virtual loss.

        Args:
            node (MCTSNode): The current node in the Monte Carlo Tree Search.
            valid_moves (np.ndarray): A boolean or integer array indicating which moves are valid (typically of length 6).

        Returns:
            Optional[int]: The index of the best action to take, or None if no valid action is found.
        """
        best_score = -float("inf")
        best_action = None

        parent_visits = max(1, node.visit_count)

        for action in range(6):
            if not valid_moves[action] or action not in node.children:
                continue

            child = node.children[action]
            score = child.ucb_score(parent_visits, self.config.mcts.c_puct)

            # Account for virtual loss
            if child.virtual_loss > 0:
                score -= child.virtual_loss

            if score > best_score:
                best_score = score
                best_action = action

        return best_action

    def _backup(self, path: list, value: float) -> None:
        """
        Propagates the simulation result back through the path of visited nodes, updating visit counts and value sums.

        Args:
            path (list): A list of (state_key, action, player) tuples representing the sequence of nodes and actions taken during the simulation.
            value (float): The simulation result to be backed up, typically from the perspective of the final player.

        Notes:
            - Increments visit counts for each node and child along the path.
            - Adjusts virtual loss for each child node.
            - Updates value sums based on whether the child node's current player matches the final player.
        """
        # Value is from the perspective of the player at the leaf
        # We need to flip it as we go up the tree
        current_value = value

        for state_key, action, player in reversed(path):
            node = self.nodes[state_key]
            child = node.children[action]

            with self.lock:
                child.visit_count += 1
                child.virtual_loss = max(0, child.virtual_loss - 1)

                # Update value from the perspective of the player who made this move
                child.value_sum += current_value
                node.visit_count += 1

            # Flip value for the opponent
            current_value = -current_value

    def _state_key(self, game: KalahGame) -> str:
        """
        Generate a unique string key representing the current state of the game.

        The key is composed of the current player's identifier and a hexadecimal
        representation of the canonical game state. This is useful for hashing or
        caching game states in algorithms such as Monte Carlo Tree Search (MCTS).

        Args:
            game (KalahGame): The current game instance.

        Returns:
            str: A unique string key for the given game state.
        """
        return f"{game.current_player}:{game.get_canonical_state().tobytes().hex()}"

    def clear_tree(self) -> None:
        """
        Clears all nodes from the MCTS tree in a thread-safe manner.

        This method acquires a lock to ensure that the operation is safe in multi-threaded environments,
        then removes all nodes from the internal node storage.
        """
        with self.lock:
            self.nodes.clear()

    def get_action_probabilities(
        self, game: KalahGame, temperature: float = 1.0
    ) -> npt.NDArray[np.float64]:
        """
        Calculates the probability distribution over actions based on visit counts from MCTS search.
        Args:
            game (KalahGame): The current game state for which to compute action probabilities.
            temperature (float, optional): Controls the level of exploration. A value close to 0 selects the most visited action deterministically, while higher values increase exploration. Defaults to 1.0.
        Returns:
            npt.NDArray[np.float64]: A probability distribution over possible actions, summing to 1.0.
        """
        visits = self.search(game)

        # Add small constant to avoid zero division
        visits = visits + 1e-8

        if temperature == 0 or temperature < 1e-8:
            probs = np.zeros_like(visits)
            probs[np.argmax(visits)] = 1.0
        else:
            visits_temp = np.power(visits, 1.0 / temperature)
            probs = visits_temp / np.sum(visits_temp)

        return probs

File: server/mcts/mcts.cpp
#include "mcts.hpp"
#include "kalah_game.hpp"
#include "neural_network.hpp"
#include <algorithm>
#include <numeric>
#include <sstream>
#include <iomanip>

// Thread-local RNG
thread_local std::mt19937 MCTS::rng_{std::random_device{}()};

// MCTSNode implementation
MCTSNode::MCTSNode(float prior) : prior(prior) {}

float MCTSNode::value() const
{
    int visits = visit_count.load(std::memory_order_acquire);
    if (visits == 0)
        return 0.0f;
    return value_sum.load(std::memory_order_acquire) / visits;
}

float MCTSNode::ucb_score(int parent_visits, float c_puct) const
{
    int visits = visit_count.load(std::memory_order_acquire);

    if (visits == 0)
    {
        return c_puct * prior * std::sqrt(static_cast<float>(parent_visits));
    }

    float exploitation = value();
    float exploration = c_puct * prior * std::sqrt(static_cast<float>(parent_visits)) / (1 + visits);
    float virtual_penalty = virtual_loss.load(std::memory_order_acquire) * c_puct / (1 + visits);

    return exploitation + exploration - virtual_penalty;
}

bool MCTSNode::is_expanded() const
{
    return expanded.load(std::memory_order_acquire);
}

bool MCTSNode::try_expand()
{
    bool expected = false;
    return expanded.compare_exchange_strong(expected, true, std::memory_order_acq_rel);
}

void MCTSNode::add_visit(float value)
{
    visit_count.fetch_add(1, std::memory_order_acq_rel);
    // Atomic float addition
    float old_sum = value_sum.load(std::memory_order_acquire);
    while (!value_sum.compare_exchange_weak(old_sum, old_sum + value,
                                            std::memory_order_acq_rel,
                                            std::memory_order_acquire))
        ;
}

void MCTSNode::add_virtual_loss()
{
    virtual_loss.fetch_add(1, std::memory_order_acq_rel);
}

void MCTSNode::remove_virtual_loss()
{
    virtual_loss.fetch_sub(1, std::memory_order_acq_rel);
}

// GPUBatchProcessor implementation
GPUBatchProcessor::GPUBatchProcessor(int gpu_id, int batch_size)
    : gpu_id_(gpu_id), batch_size_(batch_size)
{
    network_ = std::make_unique<NeuralNetwork>(gpu_id);
}

GPUBatchProcessor::~GPUBatchProcessor()
{
    stop();
}

void GPUBatchProcessor::start()
{
    running_ = true;
    worker_thread_ = std::thread(&GPUBatchProcessor::process_batches, this);
}

void GPUBatchProcessor::stop()
{
    running_ = false;
    cv_.notify_all();
    if (worker_thread_.joinable())
    {
        worker_thread_.join();
    }
}

void GPUBatchProcessor::submit_request(InferenceRequest &&request)
{
    {
        std::lock_guard<std::mutex> lock(queue_mutex_);
        request_queue_.push(std::move(request));
    }
    cv_.notify_one();
}

void GPUBatchProcessor::process_batches()
{
    std::vector<InferenceRequest> batch;
    batch.reserve(batch_size_);

    while (running_)
    {
        {
            std::unique_lock<std::mutex> lock(queue_mutex_);
            cv_.wait(lock, [this]
                     { return !request_queue_.empty() || !running_; });

            if (!running_)
                break;

            // Collect batch
            while (!request_queue_.empty() && batch.size() < batch_size_)
            {
                batch.push_back(std::move(request_queue_.front()));
                request_queue_.pop();
            }
        }

        if (!batch.empty())
        {
            // Prepare batch data
            std::vector<float> batch_states;
            batch_states.reserve(batch.size() * batch[0].state.size());

            for (const auto &req : batch)
            {
                batch_states.insert(batch_states.end(), req.state.begin(), req.state.end());
            }

            // Run inference
            auto [policies, values] = network_->predict_batch(batch_states, batch.size());

            // Distribute results
            for (size_t i = 0; i < batch.size(); ++i)
            {
                std::vector<float> policy(policies.begin() + i * 6, policies.begin() + (i + 1) * 6);
                batch[i].promise.set_value({policy, values[i]});
            }

            batch.clear();
        }
    }
}

// MCTS implementation
MCTS::MCTS(const MCTSConfig &config, int num_threads, int num_gpus)
    : config_(config), num_threads_(num_threads)
{

    // Initialize GPU processors
    for (int i = 0; i < num_gpus; ++i)
    {
        gpu_processors_.push_back(std::make_unique<GPUBatchProcessor>(i, 64));
        gpu_processors_[i]->start();
    }

    // Initialize thread pool
    for (int i = 0; i < num_threads_; ++i)
    {
        thread_pool_.emplace_back([this]
                                  {
            while (!stop_threads_) {
                std::function<void()> task;
                {
                    std::unique_lock<std::mutex> lock(task_mutex_);
                    task_cv_.wait(lock, [this] { return !task_queue_.empty() || stop_threads_; });
                    
                    if (stop_threads_) break;
                    
                    task = std::move(task_queue_.front());
                    task_queue_.pop();
                }
                task();
            } });
    }
}

MCTS::~MCTS()
{
    // Stop thread pool
    stop_threads_ = true;
    task_cv_.notify_all();
    for (auto &thread : thread_pool_)
    {
        thread.join();
    }

    // GPU processors are automatically stopped in their destructors
}

std::string MCTS::state_key(const KalahGame &game) const
{
    std::stringstream ss;
    ss << game.get_current_player() << ":";

    auto state = game.get_canonical_state();
    for (float val : state)
    {
        ss << std::setprecision(0) << std::fixed << val << ",";
    }

    return ss.str();
}

std::shared_ptr<MCTSNode> MCTS::get_or_create_node(const std::string &key)
{
    // Try to find existing node without lock
    auto it = nodes_.find(key);
    if (it != nodes_.end())
    {
        return it->second;
    }

    // Need to create - use lock
    std::lock_guard<std::mutex> lock(nodes_mutex_);

    // Double-check after acquiring lock
    it = nodes_.find(key);
    if (it != nodes_.end())
    {
        return it->second;
    }

    // Create new node
    auto node = std::make_shared<MCTSNode>();
    nodes_[key] = node;
    return node;
}

std::vector<float> MCTS::search(const KalahGame &game, const std::string &root_state_key)
{
    std::string root_key = root_state_key.empty() ? state_key(game) : root_state_key;

    // Initialize root node
    auto root_node = get_or_create_node(root_key);
    root_node->current_player = game.get_current_player();

    // Expand root if needed
    if (!root_node->is_expanded())
    {
        auto game_copy = game.clone();
        auto state = game_copy.get_canonical_state();
        auto [policy, _] = predict(state);
        expand_node(game_copy, root_key, policy);
    }

    // Add Dirichlet noise to root
    auto valid_moves = game.get_valid_moves();
    std::gamma_distribution<float> gamma(config_.dirichlet_alpha, 1.0f);
    std::vector<float> noise(6);
    float noise_sum = 0.0f;

    for (int i = 0; i < 6; ++i)
    {
        if (valid_moves[i])
        {
            noise[i] = gamma(rng_);
            noise_sum += noise[i];
        }
    }

    // Normalize noise and apply to root children
    for (int action = 0; action < 6; ++action)
    {
        if (valid_moves[action] && root_node->children.count(action))
        {
            auto child = root_node->children[action];
            child->prior = (1 - config_.dirichlet_epsilon) * child->prior +
                           config_.dirichlet_epsilon * (noise[action] / noise_sum);
        }
    }

    // Run simulations in batches
    int simulations_done = 0;
    while (simulations_done < config_.num_simulations)
    {
        int batch_size = std::min(64, config_.num_simulations - simulations_done);
        simulate_batch(batch_size, game, root_key);
        simulations_done += batch_size;
    }

    // Extract visit counts
    std::vector<float> visits(6, 0.0f);
    for (const auto &[action, child] : root_node->children)
    {
        visits[action] = static_cast<float>(child->visit_count.load(std::memory_order_acquire));
    }

    return visits;
}

void MCTS::simulate_batch(int batch_size, const KalahGame &root_game, const std::string &root_state)
{
    std::vector<std::future<void>> futures;

    for (int i = 0; i < batch_size; ++i)
    {
        auto promise = std::make_shared<std::promise<void>>();
        futures.push_back(promise->get_future());

        // Submit task to thread pool
        {
            std::lock_guard<std::mutex> lock(task_mutex_);
            task_queue_.push([this, root_game, root_state, promise]()
                             {
                auto task = simulate_to_leaf(root_game, root_state);
                
                if (task.game->is_game_over()) {
                    // Terminal node - backup immediately
                    float value = task.game->get_reward(task.game->get_current_player());
                    backup_lockfree(task.path, value);
                } else {
                    // Need neural network evaluation
                    auto state = task.game->get_canonical_state();
                    InferenceRequest request;
                    request.state = state;
                    
                    auto result_future = request.promise.get_future();
                    
                    // Submit to GPU
                    int gpu_id = next_gpu_.fetch_add(1) % gpu_processors_.size();
                    gpu_processors_[gpu_id]->submit_request(std::move(request));
                    
                    // Wait for result
                    auto [policy, value] = result_future.get();
                    
                    // Expand if needed
                    auto node = get_or_create_node(state_key(*task.game));
                    if (node->try_expand()) {
                        expand_node(*task.game, state_key(*task.game), policy);
                    }
                    
                    // Backup
                    backup_lockfree(task.path, value);
                }
                
                promise->set_value(); });
        }
        task_cv_.notify_one();
    }

    // Wait for all simulations to complete
    for (auto &future : futures)
    {
        future.wait();
    }
}

MCTS::SimulationTask MCTS::simulate_to_leaf(const KalahGame &game, const std::string &root_state)
{
    SimulationTask task;
    task.game = std::make_unique<KalahGame>(game);
    task.root_state = root_state;

    std::string current_state = root_state;

    while (true)
    {
        auto node = get_or_create_node(current_state);
        node->current_player = task.game->get_current_player();

        // Check terminal
        if (task.game->is_game_over())
        {
            return task;
        }

        // Check if needs expansion
        if (!node->is_expanded())
        {
            return task;
        }

        auto valid_moves = task.game->get_valid_moves();
        bool has_valid_moves = std::any_of(valid_moves.begin(), valid_moves.end(),
                                           [](bool v)
                                           { return v; });
        if (!has_valid_moves)
        {
            return task;
        }

        // Select action
        int action = select_action_lockfree(*node, valid_moves);
        if (action == -1)
        {
            return task;
        }

        // Add virtual loss
        auto child = node->children[action];
        child->add_virtual_loss();

        task.path.push_back({current_state, action, task.game->get_current_player()});

        // Make move
        bool extra_turn = task.game->make_move(action);
        current_state = state_key(*task.game);

        if (!extra_turn && task.game->is_game_over())
        {
            return task;
        }
    }
}

int MCTS::select_action_lockfree(const MCTSNode &node, const std::vector<bool> &valid_moves)
{
    int parent_visits = std::max(1, node.visit_count.load(std::memory_order_acquire));

    float best_score = -std::numeric_limits<float>::infinity();
    int best_action = -1;

    for (int action = 0; action < 6; ++action)
    {
        if (!valid_moves[action] || node.children.find(action) == node.children.end())
        {
            continue;
        }

        float score = node.children.at(action)->ucb_score(parent_visits, config_.c_puct);

        if (score > best_score)
        {
            best_score = score;
            best_action = action;
        }
    }

    return best_action;
}

void MCTS::expand_node(KalahGame &game, const std::string &state_key, const std::vector<float> &policy)
{
    auto node = nodes_[state_key];

    // Mask invalid actions
    auto valid_moves = game.get_valid_moves();
    std::vector<float> masked_policy(6);
    float policy_sum = 0.0f;

    for (int i = 0; i < 6; ++i)
    {
        if (valid_moves[i])
        {
            masked_policy[i] = policy[i];
            policy_sum += policy[i];
        }
    }

    // Normalize
    if (policy_sum > 0)
    {
        for (float &p : masked_policy)
        {
            p /= policy_sum;
        }
    }
    else
    {
        // Uniform over valid moves
        int valid_count = std::count(valid_moves.begin(), valid_moves.end(), true);
        for (int i = 0; i < 6; ++i)
        {
            if (valid_moves[i])
            {
                masked_policy[i] = 1.0f / valid_count;
            }
        }
    }

    // Create children
    for (int action = 0; action < 6; ++action)
    {
        if (valid_moves[action])
        {
            node->children[action] = std::make_shared<MCTSNode>(masked_policy[action]);
        }
    }
}

void MCTS::backup_lockfree(const std::vector<std::tuple<std::string, int, int>> &path, float value)
{
    float current_value = value;

    // Process path in reverse
    for (auto it = path.rbegin(); it != path.rend(); ++it)
    {
        const auto &[state_key, action, player] = *it;

        auto node = nodes_[state_key];
        auto child = node->children[action];

        // Remove virtual loss and add visit
        child->remove_virtual_loss();
        child->add_visit(current_value);
        node->add_visit(0); // Just increment visit count

        // Flip value for opponent
        current_value = -current_value;
    }
}

std::pair<std::vector<float>, float> MCTS::predict(const std::vector<float> &state)
{
    InferenceRequest request;
    request.state = state;

    auto future = request.promise.get_future();

    // Submit to next GPU in round-robin
    int gpu_id = next_gpu_.fetch_add(1) % gpu_processors_.size();
    gpu_processors_[gpu_id]->submit_request(std::move(request));

    return future.get();
}

void MCTS::clear_tree()
{
    std::lock_guard<std::mutex> lock(nodes_mutex_);
    nodes_.clear();
}

std::vector<float> MCTS::get_action_probabilities(const KalahGame &game, float temperature)
{
    auto visits = search(game);

    // Add small constant
    for (float &v : visits)
    {
        v += 1e-8f;
    }

    if (temperature < 1e-8f)
    {
        // Greedy selection
        std::vector<float> probs(6, 0.0f);
        int best_action = std::distance(visits.begin(), std::max_element(visits.begin(), visits.end()));
        probs[best_action] = 1.0f;
        return probs;
    }
    else
    {
        // Apply temperature
        std::vector<float> visits_temp(6);
        float sum = 0.0f;

        for (int i = 0; i < 6; ++i)
        {
            visits_temp[i] = std::pow(visits[i], 1.0f / temperature);
            sum += visits_temp[i];
        }

        // Normalize
        for (float &v : visits_temp)
        {
            v /= sum;
        }

        return visits_temp;
    }
}